Introduce `GLOBAL_LISTENER_BATCH` record

fabiopaini-camunda opened the pull request:
## Description
This PR introduces the new `GLOBAL_LISTENER_BATCH` record containing:
- `listenersConfigKey`: a key which can be referenced by user tasks to pin the configuration
- `taskListeners`: a list of listeners; this is implemented as a different kind of record because it will be needed in the future while implementing the OC API for global listeners (no need to add a new `ValueType` for now); each one containing:
  - `type`: String
  - `retries`: String
  - `eventTypes`: List of strings
  - `afterNonGlobal`: boolean

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #41889

fabiopaini-camunda committed:
feat: introduce new GLOBAL_LISTENERS value type

fabiopaini-camunda committed:
feat: implement GLOBAL_LISTENERS record

fabiopaini-camunda committed:
feat: make GLOBAL_LISTENERS command distributable

fabiopaini-camunda committed:
feat: add exporter support for GLOBAL_LISTENERS records

fabiopaini-camunda committed:
docs: fix references and add missing mappings in guide to create a new record type

fabiopaini-camunda committed:
test: add CompactRecordLogger support for GLOBAL_LISTENERS records

fabiopaini-camunda committed:
feat: add golden files for new GLOBAL_LISTENERS record

fabiopaini-camunda committed:
feat: add noop applier for GLOBAL_LISTENERS.CONFIGURED (needed to avoid failing tests)

mustafadagher commented on zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/globallisteners/GlobalListenerRecord.java:
```diff
@@ -0,0 +1,83 @@
+    implements GlobalListenerRecordValue {
+
+  private final StringProperty typeProp = new StringProperty("type");
+  private final StringProperty retriesProp = new StringProperty("retries");
```
🔧 also related to the previous comment, from the tests I could confirm that this is indeed the number of retries.
I suggest changing this to `IntegerProperty` and the return type in the interface to `int`

mustafadagher commented on zeebe/protocol-impl/src/test/java/io/camunda/zeebe/protocol/impl/JsonSerializableToJsonTest.java:
```diff
@@ -4040,6 +4042,55 @@ final var record = new ClockRecord().offsetBy(30);
+                    .setListenersConfigKey(1)
+                    .addTaskListener(
+                        new GlobalListenerRecord()
+                            .setType("global1")
```
❓ will we have an enum of accepted types? or what is expected to be there?

entangled90 reviewed (approved):
Thanks, I just checked the changes in `zeebe/broker`

mustafadagher commented on zeebe/protocol/src/main/java/io/camunda/zeebe/protocol/record/value/GlobalListenerRecordValue.java:
```diff
@@ -0,0 +1,33 @@
+public interface GlobalListenerRecordValue extends RecordValue {
+  String getType();
+
+  String getRetries();
```
where will this feel expression come from? and where and when will it be interpreted?

mustafadagher replied:
where will this feel expression come from? and where and when will it be interpreted?

mustafadagher replied:
>It is currently parsed as an expression in [UserTaskTransformer](https://github.com/camunda/camunda/blob/ba79acae0a568895673e05aba67a3e86255f8711/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/UserTaskTransformer.java#L356)

That brings me exactly to my point. To my knowledge: so far we only parse feel-expressions during bpmn transformation, for modeling-time listeners, this will be possible. but do we know how we will do it for ones that are created from configurations?

Also, is that a product direction we know we will need to take in the future? or is is just for extensibility?

fabiopaini-camunda replied:
@mustafadagher I understand your concerns now, thank you.
The new handling of global listeners will not pass through the transformer any more, since the configuration can change any time and the list of listeners will not be bound to the model-level `ExecutableUserTask` but to the specific user task instance.

I think there could be a way to handle expressions in any case, but since it would increase the effort and there is not a current requirement for this support (the current validation of the configuration only allows numeric values anyway) I will follow your suggestion of just handling the retries as integers.

I changed the code accordingly.

mustafadagher replied:
Thanks for the explanation @fabiopaini-camunda ! 

I haven't double checked if parsing the expression would be possible beyond the transformation phase or not. I'm just unaware of any example of that.

fabiopaini-camunda committed:
feat: make GlobalListenerRecord.retries an integer instead of a string

The "retries" property contains the number of retries which should be
attempted when executing the user task listener.

For model-level listeners this is defined as a string since it is
possible to use an expression to define the number of retries, with the
expression being parsed in the UserTaskTransformer. Since global
listeners evaluation will not pass through the transformer anymore, it
will not be possible to parse an expression defined in a global
listener. Because of this, the retries property can be defined as an
integer directly.

fabiopaini-camunda committed:
feat: rename GLOBAL_LISTENERS record to GLOBAL_LISTENER_BATCH to make its purpose clearer

Co-authored-by: Nico Korthout <korthout@users.noreply.github.com>

korthout reviewed (approved):
Nice work @fabiopaini-camunda 

🔧 I have one more suggestion

LGTM 👍

fabiopaini-camunda committed:
feat: rename GlobalListenerBatchRecord.listenersConfigKey to globalListenerBatchKey to improve clarity

fabiopaini-camunda commented on zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/globallistener/GlobalListenerBatchRecord.java:
```diff
@@ -0,0 +1,49 @@
+public final class GlobalListenerBatchRecord extends UnifiedRecordValue
+    implements GlobalListenerBatchRecordValue {
+
+  private final LongProperty listenersConfigKey = new LongProperty("listenersConfigKey", -1L);
```
@korthout I'd rather change the name of the state entities if necessary, we already changed the name of the record once and every time we do that we have to ensure everything is correct in 20+ files, pushing the completion of this PR forward again.

I think the GLOBAL_LISTENER_BATCH name we decided on is a good one, since the record basically contains a group of listeners. Currently we are using it to update the configuration all at once (with the CONFIGURE and CONFIGURED intents), but nothing is stopping us from using the same record in the future to create/delete/update multiple listeners at once with different intents.

I can change the name of the persisted entities in the other PR, but to be honest I think it could make sense to keep them as they are, even if not completely aligned with this record: this record represents a "batch" of listeners being used for something (in this case to completely replace the global configuration), what we are storing in the entities introduced in #42013 is specifically the global configuration itself.

fabiopaini-camunda committed:
docs: update documentation for the creation of new records

The maps in TypedEventRegistry and CommandApiRequestReader have been
replaced by a unified one in UnifiedRecordValue
